Update TypedData in line with SNIP-12

DelevoXDG opened the pull request:
## Describe your changes

<!-- A brief description of the changes introduced in this PR -->

This PR is part of the stack:

-- Update TypedData in line with SNIP-12 (#428) 
-- Use sealed classes for basic and preset types (#440)
-- Add separate methods for primitives (#442)
-- Nest TypedData-related test cases (#434)

### TypedData Revision 0 and Revision 1
- Support `bool` basic type
- Introduce `Domain` class; Use it for `domain` instead of `JsonObject` 
- Introduce `Revision` enum
- Remove `raw` basic type; Don't allow `merkletree` with `raw` leaves type
- Rename type-related classes
  - `TypeBase`->`Type`
  - `Type`->`StandardType`

### TypedData Revision 1
- Support new types
  - Basic types: `enum`, `string` (long string), `u128`, `i128`, `ContractAddress`, `ClassHash`, `timestamp`, `shortstring` 
    - Add `EnumType`
  - Preset types: `TokenAddress`, `NtfId`, `u256`
- ~~Rename `types`->`customTypes`; Add `types` to handle `customTypes` + `presetTypes`~~ Add `allTypes` to handle `types` + `presetTypes`
- Update type verification in line with [these suggestions](https://github.com/starknet-io/SNIPs/pull/77)
- Update internal logic in line with revision 1 standards

### Other
- Expand `HashMethod` enum to hash multiple elements


## Linked issues
<!-- Reference any GitHub issues resolved by this PR starting every line with `Closes` -->

Closes #423 

## Breaking changes

- [x] This issue contains breaking changes

<!-- List all breaking changes introduced by this issue -->

- `TypedData.domain` is now of type `TypedData.Domain` instead of `JsonObject`
- Renamed `TypedData.Type`->`TypedData.StandardType`, `TypedData.TypeBase`->`TypedData.Type`
- Removed `raw` basic type

## References
(might come in handy for reviews)

- [SNIP-12](https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-12.md)
- [Additional clarifications re. types](https://github.com/starknet-io/SNIPs/pull/77)
- starknet.js [implementation](https://github.com/starknet-io/starknet.js/pull/920)
- [Serialization of Cairo types](https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/serialization_of_Cairo_types)

DelevoXDG committed:
Fix `TransactionVersion` test cases

DelevoXDG committed:
Use `Transaction.VX_QUERY` in estimate fee tests instead of checking raw value

DelevoXDG committed:
Add `TypedDataRevision` enum

DelevoXDG committed:
Add docstring for `TypedDataRevision`

DelevoXDG committed:
Add `EnumType`

DelevoXDG committed:
Update serializer to support `EnumType`

DelevoXDG committed:
Fix `TypedDataRevision` serialization

DelevoXDG committed:
Reference SNIP-12 as a standard in TypedData docstring

DelevoXDG committed:
Add `Domain` helper class

DelevoXDG committed:
Update type verification logic in `TypedData`

- Add `revision`, `domainObjectName` helper fields to `TypedData`
- Add new restrictions on types introduced in SNIP-12
- Move versioned `reservedTypes` to companion object

DelevoXDG committed:
Merge branch 'main' into feat/snip-12

DelevoXDG committed:
Remove redundand typing

DelevoXDG committed:
Fix Domain serialization

- All fields handled as JsonPrimitives
- Allow `chainId` and `chain_id` (should be verified if `chain_id` is valid for revision 0)

DelevoXDG committed:
Fix dangling types check

- Add type specific references, arrays
- Add `extractEnumTypes` helper
- Adjust tests to not have any dangling types

DelevoXDG committed:
Update `MerkleTree` to allow `Poseidon` hashing

- Add `MerkleHashFunction` enum
- `MerkleTree` now accepts `hashFunction` as an optinal argument
- (breaking) `MerkleTree.hash` now requires `hashFunction` to be specified

DelevoXDG committed:
Minor refactor of `TypedData`

- Rename `domainObjectName` to `domainSeparatorName`
- Format `referencedTypes` logic for better readability

DelevoXDG committed:
Fix TypedDataTest

DelevoXDG committed:
Don't support `chainn_id` in `TypedData.Domain`

- Remove @JsonNames from `TypedData.Domain`
- Fix `TD_SESSION` test case in `TypedDataTest`
  - Use `chainId` instead of `chain_id` in `typed_data_session_example.json`

DelevoXDG committed:
Format

DelevoXDG committed:
Rename `MerkleHashFunction` -> `HashMethod`

- Move to package com.swmansion.starknet.crypto
- Implement using abstract function instead of stored lambda

DelevoXDG committed:
Change `domain` from `JsonObject` to `TypedData.Domain`

DelevoXDG committed:
Add `hash(List<Felt>)` method to `HashMethod`

DelevoXDG committed:
Introduce revision-based hashing in `TypedData`

- Add revision-based `hashMethod`, `hashArray()`
- Use `hashArray()` for hasing arrays
- Instantiate `MerkleTree` with `hashMethod`

DelevoXDG committed:
Move `TypedData.domainSeparatorName` to `Domain.separatorName`

DelevoXDG committed:
Replace `contains` and `containKeys` calls with `in`

DelevoXDG committed:
Add helper `String.isArray()` and `String.isEnum()` extensions

DelevoXDG committed:
Fix `HashMethod.PEDERSEN.hash`

DelevoXDG committed:
Update `TypedData.getDependencies`

- Support REV1 enum types
- Support getting dep from contains for merkletree (this should probably be verified)
- Refactor logic

DelevoXDG committed:
Fix serialization process for `enum` type

- Support revision 0 type named enum

DelevoXDG committed:
Revert "Update `TypedData.getDependencies`"

This reverts commit 51abbd9a391221839379e2a4d7c1f28ca10ec287.

DelevoXDG committed:
Update `TypedData.getDependencies`

- Support rev 1 enum types
- Refactor logic

DelevoXDG committed:
Move secondary `TypedData` constructor to the top

DelevoXDG committed:
Use isArray() instead of raw check

DelevoXDG committed:
Support preset types in `TypedData`

- [BREAKING] Rename `types` to `customTypes`
- Add `presetTypesV0`, `presetTypesV1`
- Add argument to `verifyTypes`, only verify customTypes
- Add `types` as additional helper property

DelevoXDG committed:
[BREAKING] Rename Type-related classes in `TypedData`

- Rename `TypedData.Type`->`TypedData.StandardType`
- Rename `TypedData.TypeBase`->`TypedData.Type`

DelevoXDG committed:
Commit random uncommited changes

DelevoXDG committed:
Make `TypedData.verifyTypes` argument-less again

DelevoXDG committed:
Update `encodeDependency` to escape when encoding and support enums

- Rename `escapeTypeString`->`escape`

DelevoXDG committed:
Move `TypedData.escape` to `TypedData.encodeDependency.escape`

DelevoXDG committed:
Update array handling in `TypedData.encodeValue`

DelevoXDG committed:
Support handling 'enum' basic type in `TypedData.encodeValue`

DelevoXDG committed:
Move `merkletree` handling higher in `TypedData.encodeValue`

DelevoXDG committed:
Update `string` handling in `TypedData.encodeValue`

- Support long strings (rev 1)
- Add helper extension `String.splitToShortStrings`

DelevoXDG committed:
Add support for rev 1 basic types in `TypedData.encodeValue`

- Support "bool", "u128", "i128", "ContractAddress", "ClassHash", "timestamp", "shortstring"
- Modify `TypedData.feltFromPrimitive` to handle bools

DelevoXDG committed:
Update `TypedData.getMessageHash` to support rev 1

- Use correct domain separator name based on revision

DelevoXDG committed:
Add `Felt.fromSigned` helper factory constructors

DelevoXDG committed:
Properly handle `bool` and `i128` basic types in `TypedData.encodeValue`

- Support `bool` for revision 0
- Update `Felt.feltFromPrimitive` to handle negative values

DelevoXDG committed:
Refactor `reservedTypesV0` and `reservedTypesV1` to be sets

DelevoXDG committed:
Reorder checks in `TypedData.verifyTypes()`

DelevoXDG committed:
Change `TypedDataRevision` `value` type from `Felt` to `Hex`

DelevoXDG committed:
Update reserved types tests

DelevoXDG committed:
Lazily load TD from files in TypedDataTest

- Should simplify debugging test cases that do not use said TD

DelevoXDG committed:
Refactor reserved types  in `TypedData`

- Separate `reservedTypesVX` into `basicTypesVX` and `presetTypesVX`
- Convert `basicTypesVX` and `presetTypesVX`to getters
- Add helper methods `getBasicTypes` and `getPresetTypes`
- Update `verifyTypes` to check basic type and preset types separately

DelevoXDG committed:
Make `Felt.fromSigned` methods public, add docstring, slightly refactor logic

DelevoXDG committed:
Format

DelevoXDG committed:
Move `TypedDataRevision` to `TypedData.Revision`

DelevoXDG committed:
Add remaining `InvalidTypeTest` tests; Fix `TypedData.verifyTypes()`

- Refactor tests to avoid duplication

DelevoXDG commented on lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt:
```diff
@@ -80,48 +85,135 @@ import kotlinx.serialization.json.*
+        customTypes.keys.forEach {
+            require(it.isNotEmpty()) { "Types cannot be empty." }
+            require(!it.isArray()) { "Types cannot end in *. [$it] was found." }
+            require(!it.startsWith("(") && !it.endsWith(")")) { "Types cannot be enclosed in parentheses. [$it] was found." }
```
https://github.com/starknet-io/SNIPs/pull/77/files

Not sure if this should be interpreted as "cannot be enclosed in ()" or should also apply to a singular parenthesis

THenry14 replied:
I understand it as inside `()`

DelevoXDG committed:
Make `TypedData.Domain.separatorName` internal

DelevoXDG committed:
Add missing `escape()`` call in `TypedData.encodeDependency()`

DelevoXDG committed:
Refactor tests to prepare for rev 1 tests

- Make `loadTypedData` internal
- Add `CasesRev0` helper class, move rev 0 TD imported from files there
- Add `@Nested MerkletreeTest` to gorup `merkletree` tests

DelevoXDG committed:
Add rev 0 `encode type` tests

- Make `TypedData.encodeType()` internal

DelevoXDG committed:
Move typed data test files

- Move `src/test/resources/typed-data/typed_data_*.json` to `typed_data/`

DelevoXDG committed:
Add remaining `encodeType` tests

- Add `CasesRev1` in `TypedDataTest`
- Add matching new test example files
  - typed_data_basic_types_example.json
  - typed_data_enum_example.json
  - typed_data_example.json
  - typed_data_preset_types_example.json
  - typed_data_struct_array_example.json

DelevoXDG committed:
Fix encoding of name in `TypedData.encodeDependency()`

DelevoXDG committed:
Fix `enum` handling in `TypedData.encodeDependency()`

- Fix `extractEnumTypes`, use String methods instead of regex
- Adjust `encodeDependency()`:
  - Add parentheses
  - Add comma separation

DelevoXDG committed:
Adjust `enum` handling `TypedData.verifyTypes()`

DelevoXDG committed:
Change `TypedData.stripPointer()` signature

DelevoXDG committed:
Parametrize `encode type` test

DelevoXDG committed:
Fix missing dependency test

DelevoXDG committed:
Format

DelevoXDG committed:
Add rev 1 `TypedData.getTypeHash()` tests

DelevoXDG committed:
Fix `TypedData.getMessageHash()`

- Use `hashArray()` instead of pedersen

DelevoXDG committed:
Introduce `StarknetByteArray`; fix logic

- Add `StarknetByteArray`
- Move logic for encoding long string from `TypedData.prepareLongString()` to `StarknetByteArray`
- Fix logic to properly handle short strings <31
  - Do not prepend `Felt.ZERO` to `StarknetByteArray.data`

DelevoXDG committed:
Fix and refactor `TypedData.feltFromPrimitive()`

- Boolean handled outisde of string case
- Decimals always handled in the beggining, avoid additional `.long` cast
- Throw exception on invalid primitive

DelevoXDG committed:
Change `typed_data_basic_types_example.json`

- `Json` is unable to convert large number **(investigate better solution?)**

DelevoXDG committed:
Better handle `enum` in `TypedData.encodeValue()`

- Retrieve by key, not by `.first()``

DelevoXDG committed:
Use non-docs `StarknetByteArray.fromString()` logic

- Remains to be verified if it's correct one

DelevoXDG committed:
Add remaining `getStructHash` and `getMessageHashArguments` tests

- Verified against bugfixed sn.js

DelevoXDG committed:
Simplify `enum` handling in `TypedData.encodeValue()`

DelevoXDG committed:
Move `enum` `encodeValue()` logic to a helper function

DelevoXDG committed:
Move `merkletree` `encodeValue()` logic to a helper method

DelevoXDG committed:
Refactor `encodeValue()`

- Remove newlines
- Merge handling of rev 0 basic types

DelevoXDG committed:
Drop `raw` type from `encodeValue()`; Unify and improve `merkletree` and `enum` logic

- `Context` no longer has optional values; Instead, `Context` itself is optional in `encodeValue()`
  - `requireNotNull` on `Context` in `encodeValue()` for `merkletree` and `enum`
  - `merkletree` therefore no longer supports `raw` type
  - Adjust tests to reflect that fact, add `typed_data_merkletree_felt_example.json` to test both rev 1 merkletree and raw value encoding
    - Additionally, rename `CasesRev0.TD_SESSION`-> `CasesRev0.TD_STRUCT_MERKLETREE`
- Move logic for resolving type based on context to helper `resolveType()`
  - Simplify `getMerkleTreeType()` and `getEnumVariants`
- Improve error message on `.singleOrNull()` calls
- Re-order merkle and enum helper methods

DelevoXDG committed:
Fix `TD_FELT_MERKLETREE` tests `typed_data_felt_merkletree_example.json`

DelevoXDG committed:
Add simple rev 1 test with no extra types and structs

DelevoXDG committed:
Revert "Use non-docs `StarknetByteArray.fromString()` logic"

This reverts commit 03363b3964c328e2e564cf7598daf1e69cf2154f.
It was confirmed that logic from the docs should be used:
https://github.com/starknet-io/starknet.js/issues/1003#issuecomment-1994171859

DelevoXDG committed:
Merge branch 'main' into feat/snip-12

DelevoXDG committed:
Fix `StandardAccountTest` typed data tests; Add rev 1 test; Group to `SignTypedDataTest`

DelevoXDG committed:
Add `encodeType` `merkletree` tests

DelevoXDG committed:
Format

DelevoXDG committed:
Fix `merkletree with invalid contains` test

DelevoXDG committed:
Don't nest test cases

DelevoXDG committed:
Format

DelevoXDG committed:
Require `enum` to be used only in revision 1

DelevoXDG committed:
Don't nest merkletree tests

DelevoXDG commented on lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt:
```diff
@@ -136,20 +236,32 @@ data class TypedData private constructor(
-                    toVisit.add(typeStripped)
+            params.forEach { param ->
+                val extractedTypes = when {
+                    param is EnumType -> {
+                        require(revision == Revision.V1) { "'enum' basic type is not supported in revision ${revision.value}." }
+                        listOf(param.contains)
+                    }
+                    param.type.isEnum() -> {
+                        require(revision == Revision.V1) { "Enum types are not supported in revision ${revision.value}." }
+                        extractEnumTypes(param.type)
+                    }
```
Should we have these just in case, or would it be enough for us to check `verifyTypes()`

THenry14 replied:
I'd leave them

DelevoXDG commented on lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt:
I also considered to make `TypedData` a sealed class, where there would be unified logic, while parts that are revision-specific:
- `basicTypes`, `presetTypes`
- `verifyTypes()`
- `encodeValue()`
- `getDependencies()`
- `encodeDependency()`

would be placed in their respective data classes. This would reduce `require()` checks based on revision, remove revision-specific variables and functions, and probably improve overall readability

This however has some caveats, not necessary problems, but something to take into account:
- There could be slight code duplication re. similar, but diferent logic between rev 0 and 1
- Would not be able to directly instantiate `TypedData`, only `TypedDataRevX`/`TypedDataVX`; Could use factory method though
- We previously more or less followed sn.js approach, that would move us away from it

This of course something that would likely involve moving a lot of things around, so I decided to not include this in the current PR. I wonder if this is something we want to change before releasing the SNIP support, or leave for later / leave as is for now.

WDYT?

THenry14 replied:
Imo this is a nice to have, but we probably should not focus on this atm. We can create an issue with low prio for this though

DelevoXDG replied:
#441

DelevoXDG committed:
Don't trim type in `extractEnumTypes()`

DelevoXDG committed:
Merge branch 'main' into feat/snip-12

DelevoXDG committed:
Merge branch 'main' into feat/snip-12

DelevoXDG committed:
Simplify `types` logic; Make it private

DelevoXDG committed:
Mark `revision` as `@Transient`

THenry14 commented on lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt:
```diff
@@ -80,52 +84,142 @@ import kotlinx.serialization.json.*
 @Serializable
 data class TypedData private constructor(
-    val types: Map<String, List<TypeBase>>,
+    @SerialName("types")
+    val customTypes: Map<String, List<Type>>,
+
```
I don't have a strong opinion about this, so whatever you think is best. I generally like the idea of keeping the naming as close to json as possible though. it may have been a bit less descriptive at first, but it follows closer the json which is a big plus imo.

THenry14 commented on lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt:
```diff
@@ -81,52 +84,143 @@ import kotlinx.serialization.json.*
     )
 
+    @Transient
+    private val revision = domain.revision ?: Revision.V0
```
what are the reasons to have v0 as default?

THenry14 commented on lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt:
```diff
@@ -159,34 +259,62 @@ data class TypedData private constructor(
-                }
+    private fun feltFromPrimitive(primitive: JsonPrimitive, allowSigned: Boolean = false): Felt {
+        val decimal = primitive.content.toBigIntegerOrNull()
+        decimal?.let {
+            return if (allowSigned) Felt.fromSigned(it) else Felt(it)
+        }
+        primitive.booleanOrNull?.let {
+            return if (it) Felt.ONE else Felt.ZERO
+        }
+
+        if (primitive.isString) {
+            if (primitive.content == "") {
+                return Felt.ZERO
             }
-            false -> {
-                return Felt(primitive.long)
+
+            return try {
+                Felt.fromHex(primitive.content)
+            } catch (e: Exception) {
+                Felt.fromShortString(primitive.content)
             }
         }
+
+        throw IllegalArgumentException("Unsupported primitive type: $primitive")
+    }
```
it would be a very good idea to have multiple methods per type indeed

THenry14 commented on lib/src/main/kotlin/com/swmansion/starknet/data/TypedData.kt:
```diff
@@ -276,29 +438,81 @@ data class TypedData private constructor(
+            }
+        }
+
+        private val basicTypesV0: Set<String>
+            get() = setOf("felt", "bool", "string", "selector", "merkletree", "raw")
+
+        private val basicTypesV1: Set<String>
+            get() = basicTypesV0 + setOf("enum", "u128", "i128", "ContractAddress", "ClassHash", "timestamp", "shortstring")
+
+        private val presetTypesV0: Map<String, List<Type>>
+            get() = emptyMap()
+
+        private val presetTypesV1: Map<String, List<Type>>
+            get() = mapOf(
+                "u256" to listOf(
+                    StandardType("low", "u128"),
+                    StandardType("high", "u128"),
+                ),
+                "TokenAmount" to listOf(
+                    StandardType("token_address", "ContractAddress"),
+                    StandardType("amount", "u256"),
+                ),
+                "NftId" to listOf(
+                    StandardType("collection_address", "ContractAddress"),
+                    StandardType("token_id", "u256"),
+                ),
+            )
```
I was thinking the same tbh, we could really use those. Can be done in this PR but only if made in one-two commits (to make reviewing easier). otherwise, different PR

DelevoXDG replied:
#440

DelevoXDG committed:
Improve wording in `verifyTypes()` checks

DelevoXDG committed:
Remove "raw" from `basicTypesV0`

DelevoXDG committed:
Fix tests

DelevoXDG committed:
Don't discard type names with parenthesis unless they're enclosed in parentheses from both sidews

- Only `(example_name)` is not allowed, `(example_name` and `example_name)` are now considered valid

DelevoXDG committed:
Rename `types`->`allTypes`, `customTypes`->`types`

DelevoXDG committed:
Improve error messages; Fix `merkletree with invalid context` test

DelevoXDG committed:
Format

DelevoXDG committed:
Minor `feltFromPrimitive()` refactor

THenry14 reviewed (approved):
💪

DelevoXDG committed:
Use sealed classes for basic and preset types (#440)

DelevoXDG committed:
Add separate methods for primitives (#442)

DelevoXDG committed:
Nest TypedData-related test cases (#434)

DelevoXDG committed:
Merge branch 'main' into feat/snip-12

DelevoXDG committed:
Cover extra `verifyTypes()` errors

DelevoXDG committed:
Use `BasicType` instead of raw string
